From 136a164259d4082849a9b2536bb1006e8cb31ae2 Mon Sep 17 00:00:00 2001
From: Simon McVittie <simon.mcvittie@collabora.co.uk>
Date: Tue, 8 Oct 2013 16:43:47 +0100
Subject: [PATCH 3/3] dbus-run-session: remove various extra variables from the
 environment

DBUS_SESSION_BUS_PID is not mandatory to set, but we should unset it
if present, since it points to a different session's bus. Likewise for
DBUS_SESSION_BUS_WINDOWID.

Similarly, if DBUS_STARTER_BUS_TYPE and DBUS_STARTER_ADDRESS
are set (as they would be under GNOME Terminal 3.8, see
<https://bugs.freedesktop.org/show_bug.cgi?id=63119>) then they
are likely to point to a different session's bus.

[Backported to 1.6: reimplement dbus_setenv(), use roff-format man page]

Bug: https://bugs.freedesktop.org/show_bug.cgi?id=39196
Signed-off-by: Simon McVittie <simon.mcvittie@collabora.co.uk>
Reviewed-by: Colin Walters <walters@verbum.org>
Origin: backport, commit:5ee72fe2e1e8fbb7ae09174e8df3ca28228ddd9b
---
 doc/dbus-run-session.1   |  8 ++++++++
 tools/Makefile.am        |  4 ++++
 tools/dbus-run-session.c | 43 +++++++++++++++++++++----------------------
 3 files changed, 33 insertions(+), 22 deletions(-)

diff --git a/doc/dbus-run-session.1 b/doc/dbus-run-session.1
index 8b270eb..32944d9 100644
--- a/doc/dbus-run-session.1
+++ b/doc/dbus-run-session.1
@@ -92,6 +92,14 @@ The session bus' address is made available to
 .I PROGRAM
 in the environment variable
 .BR DBUS_SESSION_BUS_ADDRESS .
+.P
+The variables
+.BR DBUS_SESSION_BUS_PID ,
+.BR DBUS_SESSION_BUS_WINDOWID ,
+.BR DBUS_STARTER_BUS_TYPE
+and
+.BR DBUS_STARTER_ADDRESS
+are removed from the environment, if present.
 .SH BUGS
 Please send bug reports to the D\-Bus mailing list or bug tracker,
 see http://www.freedesktop.org/software/dbus/
diff --git a/tools/Makefile.am b/tools/Makefile.am
index 464a805..73d95fc 100644
--- a/tools/Makefile.am
+++ b/tools/Makefile.am
@@ -47,6 +47,10 @@ dbus_launch_SOURCES=				\
 
 dbus_run_session_SOURCES =			\
 	dbus-run-session.c
+
+dbus_run_session_LDADD = \
+	$(top_builddir)/dbus/libdbus-1.la \
+	$(NULL)
 endif
 
 dbus_cleanup_sockets_SOURCES=			\
diff --git a/tools/dbus-run-session.c b/tools/dbus-run-session.c
index 4f7243f..39d8839 100644
--- a/tools/dbus-run-session.c
+++ b/tools/dbus-run-session.c
@@ -36,6 +36,21 @@
 #include <sys/wait.h>
 #include <signal.h>
 
+#include "dbus/dbus.h"
+
+/* Debian-specific backport to D-Bus 1.6: dbus_setenv is not
+ * public API in this version. Assume POSIX libc. */
+#define dbus_setenv _backported_dbus_setenv
+static dbus_bool_t
+_backported_dbus_setenv (const char *name,
+                         const char *val)
+{
+  if (val == NULL)
+    return (unsetenv (name) == 0);
+  else
+    return (setenv (name, val, TRUE) == 0);
+}
+
 #define MAX_ADDR_LEN 512
 #define PIPE_READ_END  0
 #define PIPE_WRITE_END 1
@@ -100,22 +115,6 @@ oom (void)
   exit (1);
 }
 
-static void *
-xmalloc (size_t bytes)
-{
-  void *ret;
-
-  if (bytes == 0)
-    bytes = 1;
-
-  ret = malloc (bytes);
-
-  if (ret == NULL)
-    oom ();
-
-  return ret;
-}
-
 typedef enum
 {
   READ_STATUS_OK,    /**< Read succeeded */
@@ -228,7 +227,6 @@ main (int argc, char **argv)
   int requires_arg = 0;
   pid_t bus_pid;
   pid_t app_pid;
-  char *envvar;
 
   while (i < argc)
     {
@@ -397,11 +395,12 @@ main (int argc, char **argv)
 
   close (bus_address_pipe[PIPE_READ_END]);
 
-  envvar = xmalloc (strlen ("DBUS_SESSION_BUS_ADDRESS=") +
-                    strlen (bus_address) + 1);
-  strcpy (envvar, "DBUS_SESSION_BUS_ADDRESS=");
-  strcat (envvar, bus_address);
-  putenv (envvar);
+  if (!dbus_setenv ("DBUS_SESSION_BUS_ADDRESS", bus_address) ||
+      !dbus_setenv ("DBUS_SESSION_BUS_PID", NULL) ||
+      !dbus_setenv ("DBUS_SESSION_BUS_WINDOWID", NULL) ||
+      !dbus_setenv ("DBUS_STARTER_ADDRESS", NULL) ||
+      !dbus_setenv ("DBUS_STARTER_BUS_TYPE", NULL))
+    oom ();
 
   app_pid = fork ();
 
-- 
1.8.4.rc3

